### PR TITLE
engine: Warning message for no workers

### PR DESCRIPTION
**Pull request by secant**
Added warning message for users that specify a master machine but
no worker machines. Addresses issue #407.

### CODE REVIEW

File: engine/engine.go [hunk 1]
```diff
@@ -157,6 +157,11 @@ func canBoot(machines []db.Machine) bool {
 		hasMaster = hasMaster || machine.Role == db.Master
 		hasWorker = hasWorker || machine.Role == db.Worker
 	}
+
+	if hasMaster && !hasWorker {
+		log.Warning("No worker machines were specified but a master machine was.")
```
Comment (kklin) [inline engine/engine.go]: Maybe it'd be better to log based on the return value of `canBoot` in `machineTxn`? Not sure.

**Comment by secant** [comment]
Not sure if I did it in the prettiest way 😅

File: engine/engine.go [hunk 2]
```diff
@@ -167,7 +167,13 @@ func machineTxn(view db.Database, dsl dsl.Dsl, clusterID int) error {
 
 	var dslMachines = toDBMachine(dslMachinesRaw, maxPrice)
 
-	if !canBoot(dslMachines) {
+	canBoot, hasWorker := canBoot(dslMachines)
+	if !canBoot {
+
+		if !hasWorker {
```
Comment (ejj) [inline engine/engine.go]: So I think the issue was a bit too specific.  It's not so much the fact that a master was specified without workers that's the problem.  Really, it was just that if canBoot() returns false, we do nothing without warning.  Therefore, I don't think we really need to differentiate between whether or not has Worker is true.

As a result, I'd just leave canBoot as it is, log if it's false saying something along the lines of: you're missing masters and/or workers, so we aren't booting.

File: engine/engine.go [hunk 3]
```diff
@@ -168,6 +168,7 @@ func machineTxn(view db.Database, dsl dsl.Dsl, clusterID int) error {
 	var dslMachines = toDBMachine(dslMachinesRaw, maxPrice)
 
 	if !canBoot(dslMachines) {
+		log.Warning("No worker machines and/or no master machines were specified.")
```
Comment (ejj) [inline engine/engine.go]: Sorry, I'm being super flip-floppy on this.  So the problem with this is we warn when there are neither masters nor workers.  That actually shouldn't warn as that's how you indicate that the cluster should be deleted.  What we really want to do is warn if workers are specified and masters aren't or, if masters are specified and workers aren't.

Btw, the canBoot function is useless.  Probably the easiest way to achieve this is to ditch canBoot, pull it's code into machineTxn, and then you can implement the logic.

Sorry to be so wishy-washy about this . . .

File: engine/engine.go [hunk 4]
```diff
@@ -149,26 +149,32 @@ func toDBMachine(machines []dsl.Machine, maxPrice float64) []db.Machine {
...
-	if !canBoot(dslMachines) {
+	for _, machine := range dslMachines {
+		hasMaster = hasMaster || machine.Role == db.Master
+		hasWorker = hasWorker || machine.Role == db.Worker
+	}
+
+	if !hasMaster || !hasWorker {
 		dslMachines = []db.Machine{}
+
+		if (hasMaster && !hasWorker) || (hasWorker && !hasMaster) {
```
Comment (ejj) [inline engine/engine.go]: `hasMaster != hasWorker`

**Comment by ejj** [comment]
One tiny nit.  Looks good otherwise.  Ping me when you've update and I'll merge.

**Comment by ejj** [comment]
Oh two minor nits about the commit message for future reference.

1) Please word wrap to 70 characters.  Why 70 characters for commit messages?  I have no idea ... it's the standard for email which commits are occasionally sent in.

2) Don't mention issue numbers in the commit message.  A lot of projects do this, I don't particularly like it.  No good reason.

**Pull request closed without merging**